### ISSUE TITLE
TSAN FLAKE: //test/cpp/thread_manager:thread_manager_test@poller=epoll1[k8-fastbuild]

### ISSUE THREAD

**Issue by markdroth** [grpc/grpc#20678]
https://source.cloud.google.com/results/invocations/268f930c-9a2b-4b34-9738-ef4939c0a096/targets/%2F%2Ftest%2Fcpp%2Fthread_manager:thread_manager_test@poller%3Depoll1/log

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //test/cpp/thread_manager:thread_manager_test@poller=epoll1
-----------------------------------------------------------------------------
+ export GRPC_POLL_STRATEGY=epoll1
+ shift
+ test/cpp/thread_manager/thread_manager_test
I1018 15:45:45.496955585      17 ev_epoll1_linux.cc:116]     grpc epoll fd: 3
D1018 15:45:45.497130030      17 ev_posix.cc:174]            Using polling engine: epoll1
D1018 15:45:45.498518004      17 dns_resolver_ares.cc:485]   Using ares dns resolver
D1018 15:45:45.686643240      17 thread_manager_test.cc:135] DoWork() called 33 times
==================
WARNING: ThreadSanitizer: data race (pid=17)
  Write of size 8 at 0x7b0c00004b88 by thread T8:
    #0 free /tmp/clang-build/src/compiler-rt/lib/tsan/rtl/tsan_interceptors.cc:706:3 (thread_manager_test+0x449248)
    #1 grpc::ThreadManager::CleanupCompletedThreads() /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:126:38 (liblibgrpc++_Ubase.so+0x186f28)
    #2 grpc::ThreadManager::MainWorkLoop() /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:255:3 (liblibgrpc++_Ubase.so+0x1866ff)
    #3 grpc::ThreadManager::WorkerThread::Run() /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:46:13 (liblibgrpc++_Ubase.so+0x18611b)
    #4 grpc::ThreadManager::WorkerThread::WorkerThread(grpc::ThreadManager*)::$_0::operator()(void*) const /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:36:69 (liblibgrpc++_Ubase.so+0x187544)
    #5 grpc::ThreadManager::WorkerThread::WorkerThread(grpc::ThreadManager*)::$_0::__invoke(void*) /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:36:7 (liblibgrpc++_Ubase.so+0x1874e8)
    #6 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::'lambda'(void*)::operator()(void*) const /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:139:27 (liblibgpr_Ubase.so+0x17088)
    #7 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::'lambda'(void*)::__invoke(void*) /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:109:25 (liblibgpr_Ubase.so+0x16ea8)

  Previous read of size 1 at 0x7b0c00004b88 by thread T7:
    #0 grpc::ThreadManager::WorkerThread::created() const /proc/self/cwd/./src/cpp/thread_manager/thread_manager.h:127:35 (liblibgrpc++_Ubase.so+0x187b39)
    #1 grpc::ThreadManager::MainWorkLoop() /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:185:21 (liblibgrpc++_Ubase.so+0x18646c)
    #2 grpc::ThreadManager::WorkerThread::Run() /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:46:13 (liblibgrpc++_Ubase.so+0x18611b)
    #3 grpc::ThreadManager::WorkerThread::WorkerThread(grpc::ThreadManager*)::$_0::operator()(void*) const /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:36:69 (liblibgrpc++_Ubase.so+0x187544)
    #4 grpc::ThreadManager::WorkerThread::WorkerThread(grpc::ThreadManager*)::$_0::__invoke(void*) /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:36:7 (liblibgrpc++_Ubase.so+0x1874e8)
    #5 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::'lambda'(void*)::operator()(void*) const /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:139:27 (liblibgpr_Ubase.so+0x17088)
    #6 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::'lambda'(void*)::__invoke(void*) /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:109:25 (liblibgpr_Ubase.so+0x16ea8)

  As if synchronized via sleep:
    #0 nanosleep /tmp/clang-build/src/compiler-rt/lib/tsan/rtl/tsan_interceptors.cc:361:3 (thread_manager_test+0x447ced)
    #1 gpr_sleep_until /proc/self/cwd/src/core/lib/gpr/time_posix.cc:179:17 (liblibgpr_Ubase.so+0x10b40)
    #2 grpc::ThreadManagerTest::SleepForMs(int) /proc/self/cwd/test/cpp/thread_manager/thread_manager_test.cc:80:3 (thread_manager_test+0x4b88c1)
    #3 grpc::ThreadManagerTest::DoWork(void*, bool, bool) /proc/self/cwd/test/cpp/thread_manager/thread_manager_test.cc:107:3 (thread_manager_test+0x4b8b43)
    #4 grpc::ThreadManager::MainWorkLoop() /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:209:9 (liblibgrpc++_Ubase.so+0x1865ec)
    #5 grpc::ThreadManager::WorkerThread::Run() /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:46:13 (liblibgrpc++_Ubase.so+0x18611b)
    #6 grpc::ThreadManager::WorkerThread::WorkerThread(grpc::ThreadManager*)::$_0::operator()(void*) const /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:36:69 (liblibgrpc++_Ubase.so+0x187544)
    #7 grpc::ThreadManager::WorkerThread::WorkerThread(grpc::ThreadManager*)::$_0::__invoke(void*) /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:36:7 (liblibgrpc++_Ubase.so+0x1874e8)
    #8 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::'lambda'(void*)::operator()(void*) const /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:139:27 (liblibgpr_Ubase.so+0x17088)
    #9 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::'lambda'(void*)::__invoke(void*) /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:109:25 (liblibgpr_Ubase.so+0x16ea8)

  Thread T8 'grpcpp_sync_ser' (tid=26, running) created by main thread at:
    #0 pthread_create /tmp/clang-build/src/compiler-rt/lib/tsan/rtl/tsan_interceptors.cc:967:3 (thread_manager_test+0x44a4db)
    #1 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&) /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:108:10 (liblibgpr_Ubase.so+0x16a50)
    #2 grpc_core::(anonymous namespace)::ThreadInternalsPosix* grpc_core::New<grpc_core::(anonymous namespace)::ThreadInternalsPosix, char const*&, void (*&)(void*), void*&, bool*, grpc_core::Thread::Options const&>(char const*&, void (*&)(void*), void*&, bool*&&, grpc_core::Thread::Options const&) /proc/self/cwd/./src/core/lib/gprpp/memory.h:37:18 (liblibgpr_Ubase.so+0x16448)
    #3 grpc_core::Thread::Thread(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&) /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:185:11 (liblibgpr_Ubase.so+0x161bf)
    #4 grpc::ThreadManager::WorkerThread::WorkerThread(grpc::ThreadManager*) /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:34:10 (liblibgrpc++_Ubase.so+0x185fa7)
    #5 grpc::ThreadManager::Initialize() /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:146:9 (liblibgrpc++_Ubase.so+0x18743d)
    #6 TestThreadQuota() /proc/self/cwd/test/cpp/thread_manager/thread_manager_test.cc:164:21 (thread_manager_test+0x4b9013)
    #7 main /proc/self/cwd/test/cpp/thread_manager/thread_manager_test.cc:195:3 (thread_manager_test+0x4b8cdf)

  Thread T7 'grpcpp_sync_ser' (tid=45, running) created by thread T8 at:
    #0 pthread_create /tmp/clang-build/src/compiler-rt/lib/tsan/rtl/tsan_interceptors.cc:967:3 (thread_manager_test+0x44a4db)
    #1 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&) /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:108:10 (liblibgpr_Ubase.so+0x16a50)
    #2 grpc_core::(anonymous namespace)::ThreadInternalsPosix* grpc_core::New<grpc_core::(anonymous namespace)::ThreadInternalsPosix, char const*&, void (*&)(void*), void*&, bool*, grpc_core::Thread::Options const&>(char const*&, void (*&)(void*), void*&, bool*&&, grpc_core::Thread::Options const&) /proc/self/cwd/./src/core/lib/gprpp/memory.h:37:18 (liblibgpr_Ubase.so+0x16448)
    #3 grpc_core::Thread::Thread(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&) /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:185:11 (liblibgpr_Ubase.so+0x161bf)
    #4 grpc::ThreadManager::WorkerThread::WorkerThread(grpc::ThreadManager*) /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:34:10 (liblibgrpc++_Ubase.so+0x185fa7)
    #5 grpc::ThreadManager::MainWorkLoop() /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:184:35 (liblibgrpc++_Ubase.so+0x186453)
    #6 grpc::ThreadManager::WorkerThread::Run() /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:46:13 (liblibgrpc++_Ubase.so+0x18611b)
    #7 grpc::ThreadManager::WorkerThread::WorkerThread(grpc::ThreadManager*)::$_0::operator()(void*) const /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:36:69 (liblibgrpc++_Ubase.so+0x187544)
    #8 grpc::ThreadManager::WorkerThread::WorkerThread(grpc::ThreadManager*)::$_0::__invoke(void*) /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:36:7 (liblibgrpc++_Ubase.so+0x1874e8)
    #9 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::'lambda'(void*)::operator()(void*) const /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:139:27 (liblibgpr_Ubase.so+0x17088)
    #10 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::'lambda'(void*)::__invoke(void*) /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:109:25 (liblibgpr_Ubase.so+0x16ea8)

SUMMARY: ThreadSanitizer: data race /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:126:38 in grpc::ThreadManager::CleanupCompletedThreads()
==================
```

**Comment by markdroth** [comment]
Also seeing a similar failure in client_lb_end2end_test:

https://source.cloud.google.com/results/invocations/f58e2c32-8df5-4206-a774-4f1aa5578765/targets/%2F%2Ftest%2Fcpp%2Fend2end:client_lb_end2end_test@poller%3Dpoll/log

```
==================
WARNING: ThreadSanitizer: heap-use-after-free (pid=18)
  Read of size 1 at 0x7b0c00008f68 by thread T18:
    #0 grpc::ThreadManager::WorkerThread::created() const /proc/self/cwd/./src/cpp/thread_manager/thread_manager.h:127:35 (liblibgrpc++_Ubase.so+0x1872c9)
    #1 grpc::ThreadManager::MainWorkLoop() /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:185:21 (liblibgrpc++_Ubase.so+0x185bfc)
    #2 grpc::ThreadManager::WorkerThread::Run() /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:46:13 (liblibgrpc++_Ubase.so+0x1858ab)
    #3 grpc::ThreadManager::WorkerThread::WorkerThread(grpc::ThreadManager*)::$_0::operator()(void*) const /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:36:69 (liblibgrpc++_Ubase.so+0x186cd4)
    #4 grpc::ThreadManager::WorkerThread::WorkerThread(grpc::ThreadManager*)::$_0::__invoke(void*) /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:36:7 (liblibgrpc++_Ubase.so+0x186c78)
    #5 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::'lambda'(void*)::operator()(void*) const /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:139:27 (liblibgpr_Ubase.so+0x17088)
    #6 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::'lambda'(void*)::__invoke(void*) /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:109:25 (liblibgpr_Ubase.so+0x16ea8)

  Previous write of size 8 at 0x7b0c00008f68 by thread T11:
    #0 free /tmp/clang-build/src/compiler-rt/lib/tsan/rtl/tsan_interceptors.cc:706:3 (client_lb_end2end_test+0x474108)
    #1 grpc::ThreadManager::CleanupCompletedThreads() /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:126:38 (liblibgrpc++_Ubase.so+0x1866b8)
    #2 grpc::ThreadManager::MainWorkLoop() /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:255:3 (liblibgrpc++_Ubase.so+0x185e8f)
    #3 grpc::ThreadManager::WorkerThread::Run() /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:46:13 (liblibgrpc++_Ubase.so+0x1858ab)
    #4 grpc::ThreadManager::WorkerThread::WorkerThread(grpc::ThreadManager*)::$_0::operator()(void*) const /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:36:69 (liblibgrpc++_Ubase.so+0x186cd4)
    #5 grpc::ThreadManager::WorkerThread::WorkerThread(grpc::ThreadManager*)::$_0::__invoke(void*) /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:36:7 (liblibgrpc++_Ubase.so+0x186c78)
    #6 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::'lambda'(void*)::operator()(void*) const /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:139:27 (liblibgpr_Ubase.so+0x17088)
    #7 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::'lambda'(void*)::__invoke(void*) /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:109:25 (liblibgpr_Ubase.so+0x16ea8)

  Thread T18 'grpcpp_sync_ser' (tid=165, running) created by thread T11 at:
    #0 pthread_create /tmp/clang-build/src/compiler-rt/lib/tsan/rtl/tsan_interceptors.cc:967:3 (client_lb_end2end_test+0x47539b)
    #1 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&) /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:108:10 (liblibgpr_Ubase.so+0x16a50)
    #2 grpc_core::(anonymous namespace)::ThreadInternalsPosix* grpc_core::New<grpc_core::(anonymous namespace)::ThreadInternalsPosix, char const*&, void (*&)(void*), void*&, bool*, grpc_core::Thread::Options const&>(char const*&, void (*&)(void*), void*&, bool*&&, grpc_core::Thread::Options const&) /proc/self/cwd/./src/core/lib/gprpp/memory.h:37:18 (liblibgpr_Ubase.so+0x16448)
    #3 grpc_core::Thread::Thread(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&) /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:185:11 (liblibgpr_Ubase.so+0x161bf)
    #4 grpc::ThreadManager::WorkerThread::WorkerThread(grpc::ThreadManager*) /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:34:10 (liblibgrpc++_Ubase.so+0x185737)
    #5 grpc::ThreadManager::MainWorkLoop() /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:184:35 (liblibgrpc++_Ubase.so+0x185be3)
    #6 grpc::ThreadManager::WorkerThread::Run() /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:46:13 (liblibgrpc++_Ubase.so+0x1858ab)
    #7 grpc::ThreadManager::WorkerThread::WorkerThread(grpc::ThreadManager*)::$_0::operator()(void*) const /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:36:69 (liblibgrpc++_Ubase.so+0x186cd4)
    #8 grpc::ThreadManager::WorkerThread::WorkerThread(grpc::ThreadManager*)::$_0::__invoke(void*) /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:36:7 (liblibgrpc++_Ubase.so+0x186c78)
    #9 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::'lambda'(void*)::operator()(void*) const /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:139:27 (liblibgpr_Ubase.so+0x17088)
    #10 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&)::'lambda'(void*)::__invoke(void*) /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:109:25 (liblibgpr_Ubase.so+0x16ea8)

  Thread T11 'grpcpp_sync_ser' (tid=163, finished) created by thread T8 at:
    #0 pthread_create /tmp/clang-build/src/compiler-rt/lib/tsan/rtl/tsan_interceptors.cc:967:3 (client_lb_end2end_test+0x47539b)
    #1 grpc_core::(anonymous namespace)::ThreadInternalsPosix::ThreadInternalsPosix(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&) /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:108:10 (liblibgpr_Ubase.so+0x16a50)
    #2 grpc_core::(anonymous namespace)::ThreadInternalsPosix* grpc_core::New<grpc_core::(anonymous namespace)::ThreadInternalsPosix, char const*&, void (*&)(void*), void*&, bool*, grpc_core::Thread::Options const&>(char const*&, void (*&)(void*), void*&, bool*&&, grpc_core::Thread::Options const&) /proc/self/cwd/./src/core/lib/gprpp/memory.h:37:18 (liblibgpr_Ubase.so+0x16448)
    #3 grpc_core::Thread::Thread(char const*, void (*)(void*), void*, bool*, grpc_core::Thread::Options const&) /proc/self/cwd/src/core/lib/gprpp/thd_posix.cc:185:11 (liblibgpr_Ubase.so+0x161bf)
    #4 grpc::ThreadManager::WorkerThread::WorkerThread(grpc::ThreadManager*) /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:34:10 (liblibgrpc++_Ubase.so+0x185737)
    #5 grpc::ThreadManager::Initialize() /proc/self/cwd/src/cpp/thread_manager/thread_manager.cc:146:9 (liblibgrpc++_Ubase.so+0x186bcd)
    #6 grpc_impl::Server::SyncRequestThreadManager::Start() /proc/self/cwd/src/cpp/server/server_cc.cc:925:7 (liblibgrpc++_Ubase.so+0x16420e)
    #7 grpc_impl::Server::Start(grpc_impl::ServerCompletionQueue**, unsigned long) /proc/self/cwd/src/cpp/server/server_cc.cc:1240:12 (liblibgrpc++_Ubase.so+0x15e736)
    #8 grpc_impl::ServerBuilder::BuildAndStart() /proc/self/cwd/src/cpp/server/server_builder.cc:398:11 (liblibgrpc++_Ubase.so+0x14426b)
    #9 grpc::testing::(anonymous namespace)::ClientLbEnd2endTest::ServerData::Serve(std::string const&, grpc::internal::Mutex*, grpc::internal::CondVar*) /proc/self/cwd/test/cpp/end2end/client_lb_end2end_test.cc:366:25 (client_lb_end2end_test+0x4ea4c7)
    #10 void std::_Mem_fn<void (grpc::testing::(anonymous namespace)::ClientLbEnd2endTest::ServerData::*)(std::string const&, grpc::internal::Mutex*, grpc::internal::CondVar*)>::operator()<std::string&, grpc::internal::Mutex*&, grpc::internal::CondVar*&, void>(grpc::testing::(anonymous namespace)::ClientLbEnd2endTest::ServerData*, std::string&, grpc::internal::Mutex*&, grpc::internal::CondVar*&) const /usr/lib/gcc/x86_64-linux-gnu/4.9/../../../../include/c++/4.9/functional:569:11 (client_lb_end2end_test+0x4ec8a4)
    #11 void std::_Bind<std::_Mem_fn<void (grpc::testing::(anonymous namespace)::ClientLbEnd2endTest::ServerData::*)(std::string const&, grpc::internal::Mutex*, grpc::internal::CondVar*)> (grpc::testing::(anonymous namespace)::ClientLbEnd2endTest::ServerData*, std::string, grpc::internal::Mutex*, grpc::internal::CondVar*)>::__call<void, 0ul, 1ul, 2ul, 3ul>(std::tuple<>&&, std::_Index_tuple<0ul, 1ul, 2ul, 3ul>) /usr/lib/gcc/x86_64-linux-gnu/4.9/../../../../include/c++/4.9/functional:1263:11 (client_lb_end2end_test+0x4ec6fb)
    #12 void std::_Bind<std::_Mem_fn<void (grpc::testing::(anonymous namespace)::ClientLbEnd2endTest::ServerData::*)(std::string const&, grpc::internal::Mutex*, grpc::internal::CondVar*)> (grpc::testing::(anonymous namespace)::ClientLbEnd2endTest::ServerData*, std::string, grpc::internal::Mutex*, grpc::internal::CondVar*)>::operator()<void>() /usr/lib/gcc/x86_64-linux-gnu/4.9/../../../../include/c++/4.9/functional:1321:17 (client_lb_end2end_test+0x4ec5b9)
    #13 void std::_Bind_simple<std::_Bind<std::_Mem_fn<void (grpc::testing::(anonymous namespace)::ClientLbEnd2endTest::ServerData::*)(std::string const&, grpc::internal::Mutex*, grpc::internal::CondVar*)> (grpc::testing::(anonymous namespace)::ClientLbEnd2endTest::ServerData*, std::string, grpc::internal::Mutex*, grpc::internal::CondVar*)> ()>::_M_invoke<>(std::_Index_tuple<>) /usr/lib/gcc/x86_64-linux-gnu/4.9/../../../../include/c++/4.9/functional:1699:18 (client_lb_end2end_test+0x4ec518)
    #14 std::_Bind_simple<std::_Bind<std::_Mem_fn<void (grpc::testing::(anonymous namespace)::ClientLbEnd2endTest::ServerData::*)(std::string const&, grpc::internal::Mutex*, grpc::internal::CondVar*)> (grpc::testing::(anonymous namespace)::ClientLbEnd2endTest::ServerData*, std::string, grpc::internal::Mutex*, grpc::internal::CondVar*)> ()>::operator()() /usr/lib/gcc/x86_64-linux-gnu/4.9/../../../../include/c++/4.9/functional:1688:16 (client_lb_end2end_test+0x4ec4b8)
    #15 std::thread::_Impl<std::_Bind_simple<std::_Bind<std::_Mem_fn<void (grpc::testing::(anonymous namespace)::ClientLbEnd2endTest::ServerData::*)(std::string const&, grpc::internal::Mutex*, grpc::internal::CondVar*)> (grpc::testing::(anonymous namespace)::ClientLbEnd2endTest::ServerData*, std::string, grpc::internal::Mutex*, grpc::internal::CondVar*)> ()> >::_M_run() /usr/lib/gcc/x86_64-linux-gnu/4.9/../../../../include/c++/4.9/thread:115:13 (client_lb_end2end_test+0x4ec0af)
    #16 <null> <null> (libstdc++.so.6+0xb8c7f)

SUMMARY: ThreadSanitizer: heap-use-after-free /proc/self/cwd/./src/cpp/thread_manager/thread_manager.h:127:35 in grpc::ThreadManager::WorkerThread::created() const
==================
```

**Comment by vjpai** [comment]
Fixed by #20808
